src/http_server.c

```diff
@@ -0,0 +1,216 @@
+	const char *reply = "HTTP/1.1 200 OK\r\n"
+			"Content-Type: text/plain\r\n"
+			"Content-Length: %lu\r\n"
+			"Connection: close\r\n\r\n"
+			"%s"; // User readable message
+
+	int requiredSize = snprintf(NULL, 0, reply, strlen(userMessage), userMessage) + 1;
+	char *buffer = alloca(requiredSize);
+	snprintf(buffer, requiredSize, reply, strlen(userMessage), userMessage);
+
+	ssize_t write_result = write(socket, buffer, strlen(buffer));
+	(void)write_result;
+}
+
+static int process_request(int new_socket, char *id) {
+	log_info("Processing HTTP SAML request\n");
+
+	int flag = 1;
+	setsockopt(new_socket, IPPROTO_TCP, TCP_NODELAY, &flag, sizeof(int));
+
+	// Read the request
+	char request[1024];
+	ssize_t read_result = read(new_socket, request, sizeof(request));
+
+	// Check for '=id' in the response
+	// If the recevied request from the server is larger than the buffer, the result will not be null-terminated.
+	// Causing strlen to behave wrong.
+	if (read_result < 0) {
+		log_error("Bad request\n");
+		send_status_response(new_socket, "Invalid redirect response from Fortinet server. VPN could not be established.");
+		return -1;
+	}
+
+	// Safety Null-terminate
+	request[sizeof(request) - 1] = 0;
+	request[read_result - 1] = 0;
+
+	static const char *request_head = "GET /?id=";
+
+	if (strncmp(request, request_head, strlen(request_head)) != 0) {
+		log_error("Bad request\n");
+		send_status_response(new_socket, "Invalid redirect response from Fortinet server. VPN could not be established.");
+		return -1;
+	}
+
+	// Extract the id
+	static const char *token_delimiter = " &\r\n";
+	char *next_token = request + strlen(request_head); // strsep does modify the input argument and we don't want to loose our request pointer.
+	char *id_start = strsep(&next_token, token_delimiter);
+
+	if (next_token == NULL) {
+		// In case not found, next_token was set to NULL
+		// This should be invalid because we expect \r\n at the end of the GET request line
+		log_error("Bad request format\n");
+		send_status_response(new_socket, "Invalid formatting of Fortinet server redirect response. VPN could not be established.");
+		return -1;
+	}
+
+	// strsep inserted a NULL at the location of the delimiter.
+	int id_length = strlen(id_start);
+
+	if(id_length == 0 || id_length >= MAX_SAML_SESSION_ID_LENGTH) {
+		log_error("Bad request id\n");
+		send_status_response(new_socket, "Invalid SAML session id received from Fortinet server. VPN could not be established.");
+		return -1;
+	}
+
+	// It was checked above, that the length is smaller than MAX_SAML_SESSION_ID_LENGTH
+	strcpy(id, id_start);
+
+	for (int i = 0; i < id_length; i++) {
+		if (isalnum(id[i]) || id[i] == '-') continue;
+		log_error("Invalid id format\n");
+		send_status_response(new_socket, "Invalid SAML session id received from Fortinet server. VPN could not be established.");
+		return -1;
+	}
+
+	log_info("Extracted id: %s\n", id);
```

Comment:
I am wondering if the ID should be kept secret. With this is end up in the terminal and logs, where other tools with access could possibly extract it.

Reply:
How long is the ID valid? If a new one is needed to reconnect, then I think it doesn't hurt to keep the log, but demote it to debug. If it's usable more than once, then yes, it probably shouldn't be logged even in debug.

Reply:
I have no idea how long they are valid and I don't want to mess with our IT department to find out.
Anyways it would be a race condition with a malicious tool to extract the id faster from the logs than it is being used by us.
I've decided to remove this message completely.